src/tracer.js

```diff
@@ -0,0 +1,255 @@
+    *        be unspecified.
+    * @param {array} [fields.references] - an array of Reference instances,
+    *        each pointing to a causal parent SpanContext. If specified,
+    *        `fields.childOf` must be unspecified.
+    * @param {object} [fields.tags] - set of key-value pairs which will be set
+    *        as tags on the newly created Span. Ownership of the object is
+    *        passed to the created span for efficiency reasons (the caller
+    *        should not modify this object after calling startSpan).
+    * @param {number} [fields.startTime] - a manually specified start time for
+    *        the created Span object. The time should be specified in
+    *        milliseconds as Unix timestamp. Decimal value are supported
+    *        to represent time values with sub-millisecond accuracy.
+    * @return {Span} - a new Span object.
+    **/
+    startSpan(fields: startSpanArgs): Span {
+        let tags = fields.tags || {};
+        let references = fields.references || [];
+        let startTime = fields.startTime;
+        if (!startTime) {
+            startTime = Utils.getTimestampMicros();
+        }
+
+        let parent: ?SpanContext = fields.childOf;
+        if (!parent) {
+            // If there is no child of set, then search list of references
+            for (let i = 0; i < references.length; i++) {
+                let ref: Reference = references[i];
+                let ref_type = ref.type();
+                if ((ref_type === opentracing.REFERENCE_CHILD_OF) ||
+                    (ref_type === opentracing.REFERENCE_FOLLOWS_FROM)) {
+                    parent = ref.referencedContext();
+                    break;
+                } else {
+                    // TODO(oibe) support other types of span references
+                }
+            }
+        }
+
+        let spanKindValue = tags[opentracing_tags.SPAN_KIND];
+        let rpcServer = (spanKindValue === opentracing_tags.SPAN_KIND_RPC_SERVER);
+
+
+        // $FlowIgnore - I just want a span context up front.
+        let ctx: SpanContext = new SpanContext();
+        if (!parent) {
+            let randomId = Utils.getRandom64();
+            let flags = 0;
+            if (this._sampler.isSampled()) {
+                flags |= constants.SAMPLED_MASK;
+            }
+
+            ctx.traceId = randomId;
+            ctx.spanId = randomId;
+            ctx.parentId = null;
+            ctx.flags = flags;
+        } else {
+            ctx.traceId = parent.traceId;
+            if (rpcServer) {
+                // Support Zipkin's one-span-per-RPC model
+                ctx.spanId = parent.spanId;
+                ctx.parentId = parent.parentId;
+            } else {
+                ctx.spanId = Utils.getRandom64();
+                ctx.parentId = parent.spanId;
+            }
+            ctx.flags = parent.flags;
+
+            // copy baggage items
+            ctx.baggage = _.assign(parent.baggage);
```

Comment:
we don't care if someone modifies the fields, it's their fault. The design is that the SpanContext contains a snapshot of baggage to which the tracer will never add anything, so we do not need to make copies of that map when creating children spans, we can point to the same map.